### PR TITLE
Python: Fix django taint sinks

### PR DESCRIPTION
**Pull request by RasmusWL**
I also did some major restructuring of the code. A bit controversial since it renames/moves classes that are already public. We should figure out if that is OK before we merge this PR.

Fixes https://github.com/github/codeql-python-team/issues/38
Fixes https://github.com/github/codeql/issues/3466

### CODE REVIEW

**Comment by RasmusWL** [comment]
Started https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/95/ to see if there are any differences on django/django

**Comment by RasmusWL** [comment]
Not including the stack-trace exposure in https://github.com/django/django/blob/c2250cfb80e27cdf8d098428824da2800a18cadf/django/core/handlers/asgi.py#L217-L220 does not look good. I think we will need to start differentiating XSS sinks and what is included in a HTTP response. (A JsonResponse could contain a stack-trace for sure)

**Comment by RasmusWL** [comment]
The *proper* solution to this problem of not all HttpResponses being vulnerable to XSS is probably to define a new abstract class in Http.qll called HttpResponseXSSVulnerableSink (or similar). I would like to model a few more libraries/frameworks before fully comitting to an approach though.

File: python/ql/src/semmle/python/web/django/Response.qll [hunk 1]
```diff
@@ -4,38 +4,20 @@ import semmle.python.security.strings.Basic
...
-    ) and
-    // TODO: does this do anything? when could they be the same???
-    not result = theDjangoHttpRedirectClass()
+/** INTERNAL class used for tracking a django response object. */
+private class DjangoResponseKind extends TaintKind {
+    DjangoResponseKind() { this = "django.response.HttpResponse" }
 }
 
-/** internal class used for tracking a django response. */
+/** INTENRAL taint-source used for tracking a django response. */
```
Comment (tausbn) [inline python/ql/src/semmle/python/web/django/Response.qll:12]:
```suggestion
/** INTERNAL taint-source used for tracking a django response object. */
```

File: python/ql/src/semmle/python/web/django/Response.qll [hunk 2]
```diff
@@ -55,27 +37,42 @@ class DjangoResponseWrite extends HttpResponseTaintSink {
...
+ */
+class DjangoResponseContentXSSVulnerable extends DjangoResponseContent {
+    override DjangoXSSVulnerableResponseClass cls;
+
+    DjangoResponseContentXSSVulnerable() {
+        not exists(cls.getContentTypeArg(call))
+        or
+        exists(StringValue s |
+            cls.getContentTypeArg(call).pointsTo(s) and
+            s.getText().indexOf("text/html") = 0
```
Comment (tausbn) [inline python/ql/src/semmle/python/web/django/Response.qll:68]: Would using `matches` be better here? The micro-optimiser inside me can't help but notice that `indexOf` will (potentially) run through the entire string if the prefix fails to be `text/html`.

File: python/ql/src/semmle/python/web/django/Shared.qll [hunk 3]
```diff
@@ -1,12 +1,73 @@
...
+            base = Value::named("django.http.response.HttpResponse")
+            or
+            // version 2.x and 3.x
+            // https://docs.djangoproject.com/en/2.2/ref/request-response/#httpresponse-objects
+            base = Value::named("django.http.HttpResponse")
+        ) and
+        this.getASuperType() = base
+    }
+
+    // The reason these two method are defined in this class (and not in the Sink
```
Comment (tausbn) [inline python/ql/src/semmle/python/web/django/Shared.qll:37]:
```suggestion
    // The reason these two methods are defined in this class (and not in the Sink
```

File: python/ql/src/semmle/python/web/django/Shared.qll [hunk 4]
```diff
@@ -1,12 +1,73 @@
...
+    // definition.
+
+    /** Gets the `content` argument of a `call` to the constructor */
+    ControlFlowNode getContentArg(CallNode call) { none() }
+
+    /** Gets the `content_type` argument of a `call` to the constructor */
+    ControlFlowNode getContentTypeArg(CallNode call) { none() }
+}
+
+/** A Class that is a Django Response, and is vulnerable to XSS. */
```
Comment (tausbn) [inline python/ql/src/semmle/python/web/django/Shared.qll:50]:
```suggestion
/** A class that is a Django Response, and is vulnerable to XSS. */
```

File: python/ql/src/semmle/python/web/django/Response.qll (hunk 2 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/web/django/Response.qll]: yes, excellent idea :+1:

**Comment by tausbn** [comment]
Very nice! Merging. :+1:

**Pull request merged**